### PR TITLE
COMPAT/REF: Use s3fs for s3 IO

### LINKED ISSUE #11915: io/common.py: boto3 with python 3.5
Pandas v0.17.1 won't import for Python 3.5 due to boto. Replacing with boto3 appears to fix the issue. I'd like to suggest replacing dependencies on boto with boto3 for at least Python 3.5. I didn't test with other Python 3x versions. Thank you

Update (2016-01-10T02:45:00Z): There are significant API changes between `boto` and `boto3` (http://boto3.readthedocs.org/en/latest/guide/migrations3.html). Doing `import boto3 as boto` as below will allow pandas to import for Python 3.5, but then AWS functionality is broken.

```
samuel_harrold@instance-20151227t225000z:~$ ipython
Python 3.5.1 |Anaconda 2.4.1 (64-bit)| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import pandas
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-d6ac987968b6> in <module>()
----> 1 import pandas

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/__init__.py in <module>()
     40 
     41 # let init-time option registration happen
---> 42 import pandas.core.config_init
     43 
     44 from pandas.core.api import *

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/core/config_init.py in <module>()
     15                                 is_instance_factory, is_one_of_factory,
     16                                 get_default_val)
---> 17 from pandas.core.format import detect_console_encoding
     18 
     19 

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/core/format.py in <module>()
      8 from pandas.core.base import PandasObject
      9 from pandas.core.common import adjoin, notnull
---> 10 from pandas.core.index import Index, MultiIndex, _ensure_index
     11 from pandas import compat
     12 from pandas.compat import(StringIO, lzip, range, map, zip, reduce, u,

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/core/index.py in <module>()
     29 from pandas.core.strings import StringAccessorMixin
     30 from pandas.core.config import get_option
---> 31 from pandas.io.common import PerformanceWarning
     32 
     33 

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/io/common.py in <module>()
     66 
     67 try:
---> 68     from boto.s3 import key
     69     class BotoFileLikeReader(key.Key):
     70         """boto Key modified to be more file-like

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/boto/__init__.py in <module>()
   1214     return storage_uri(uri_str)
   1215 
-> 1216 boto.plugin.load_plugins(config)

/home/samuel_harrold/anaconda3/lib/python3.5/site-packages/boto/plugin.py in load_plugins(config)
     90         return
     91     directory = config.get('Plugin', 'plugin_directory')
---> 92     for file in glob.glob(os.path.join(directory, '*.py')):
     93         _import_module(file)

/home/samuel_harrold/anaconda3/lib/python3.5/posixpath.py in join(a, *p)
     87                 path += sep + b
     88     except (TypeError, AttributeError, BytesWarning):
---> 89         genericpath._check_arg_types('join', a, *p)
     90         raise
     91     return path

/home/samuel_harrold/anaconda3/lib/python3.5/genericpath.py in _check_arg_types(funcname, *args)
    141         else:
    142             raise TypeError('%s() argument must be str or bytes, not %r' %
--> 143                             (funcname, s.__class__.__name__)) from None
    144     if hasstr and hasbytes:
    145         raise TypeError("Can't mix strings and bytes in path components") from None

TypeError: join() argument must be str or bytes, not 'NoneType'

In [2]: exit()
samuel_harrold@instance-20151227t225000z:~$ cp /home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/io/common_boto3.py /home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/io/common.py
samuel_harrold@instance-20151227t225000z:~$ diff /home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/io/common.py /home/samuel_harrold/anaconda3/lib/python3.5/site-packages/pandas/io/common_orig.py
68c68
<     from boto3.s3 import key

---
>     from boto.s3 import key
272c272
<             import boto3 as boto

---
>             import boto
samuel_harrold@instance-20151227t225000z:~$ ipython
Python 3.5.1 |Anaconda 2.4.1 (64-bit)| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import pandas

In [2]: # success

In [3]: exit()
samuel_harrold@instance-20151227t225000z:~$ ipython
Python 3.5.1 |Anaconda 2.4.1 (64-bit)| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import pandas as pd

In [2]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.16.0-0.bpo.4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 7.1.2
setuptools: 19.1.1
Cython: 0.23.4
numpy: 1.10.2
scipy: 0.16.0
statsmodels: 0.6.1
IPython: 4.0.1
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.4
matplotlib: 1.5.0
openpyxl: 2.2.6
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.7.7
lxml: 3.4.4
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.9
pymysql: None
psycopg2: None
Jinja2: None

In [3]: 
```

### PR DESCRIPTION
**Pull request by TomAugspurger**
- [x] closes #11915
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -320,31 +241,21 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
...
-            filepath_or_buffer = BytesIO(k.get_contents_as_string(
-                encoding=encoding))
-        else:
-            k = BotoFileLikeReader(b, parsed_url.path, encoding=encoding)
-            k.open('r')  # Expose read errors immediately
-            filepath_or_buffer = k
-        return filepath_or_buffer, None, compression
+            return filepath_or_buffer, encoding, compression
+        except FileNotFoundError:
+            raise boto.exception.S3ResponseError
```
Comment (jreback) [inline pandas/io/common.py]: is this an appropriate error? why is ``FileNotFoundError`` ok here?

Comment (jreback) [inline pandas/io/common.py]: actually is this try/except needed? (as its not around ``fs.open`` anyhow)

**Comment by jreback** [comment]
need to add ``s3fs`` to ``.pip`` (where we have ``boto`` now)

**Comment by TomAugspurger** [comment]
This is just refactoring the reading for now. We should be able to support writing to s3 buckets pretty easily now too. A few things

- pandas accepts `s3://`, `s3n://`, and `s3a://` protocols, but s3fs only expects the regular s3://. I work around that [here](https://github.com/pydata/pandas/pull/13137/files#diff-0d7b5a2c72b4dfc11d80afe159d45ff8R251). [This](https://wiki.apache.org/hadoop/AmazonS3) was the best source I found on the differences, but I don't know if simply chaning the protocol will break anything. At the end of the day we're just pulling down bytes, which should presumably represent a CSV
- API Changes: Before we raised OSError when the s3 file wasn't found, and reraised botos PermissionError when you didn't have permission to read the file. I think s3fs exposes both of these as IOErrors [here](https://github.com/dask/s3fs/blob/master/s3fs/core.py#L631). We might be able to work around this

cc @jreback @martindurant

File: pandas/io/common.py (hunk 1 above)
Comment (TomAugspurger) [inline pandas/io/common.py]: Not sure why I added that... Doesn't seem to be necessary.

**Comment by martindurant** [comment]
s3fs now includes original boto3 exceptions when something goes wrong https://github.com/dask/s3fs/pull/42

For your question above, I believe s3: and s3n: work exactly the same from our point of view; but we don't touch the block filesystem emulation (this is ancient, I don't know if it's used in the wild).

The failure above appears to be a typo:
```
File "/home/travis/build/pydata/pandas/pandas/io/parsers.py", line 36, in <module>
    need_text_wrapping = (compat.BtyesIO,)
AttributeError: module 'pandas.compat' has no attribute 'BtyesIO'
```

**Comment by jseabold** [comment]
I haven't yet tried this PR, but I've had some issues with s3fs lately [1], and I'd like to try this out before it gets merged. Assuming liberal permissions on s3 buckets and objects and dealing with profiles has been a bit of an issue across third-party tools for me lately. It could be an issue with permissions, but I suspect it's going to continue to come up.

E.g., I can do this

```
import boto3
session = boto3.Session(profile_name='myprofile')
s3 = session.client('s3')
key = s3.get_object(Bucket=BUCKET, Key=KEY)
```

but not 

```
import s3fs
fs = s3fs.S3FileSystem(profile_name='myprofile')
f = fs.open('s3://' + BUCKET + '/' + KEY)
```

and I don't see a way to set the profile explicitly in S3File and it doesn't seem to matter if I set my `AWS_PROFILE` correctly, but I haven't gone deep here.

[1] https://github.com/dask/s3fs/issues/38

**Comment by TomAugspurger** [comment]
@jseabold yeah, I've been hitting the same issue. For your specific case, maybe try

```python
import s3fs
fs = s3fs.S3FileSystem(profile_name='myprofile', anon=False)
f = fs.open('s3://' + BUCKET + '/' + KEY)
```

If I use anon=False though, I'm unable to read from publicly available files like `s3://pandas-test/tips.csv'

I spent last night trying to figure out a way to handle both, no luck yet.

**Comment by jseabold** [comment]
Ah, sure. Isn't there a try/except for that in the boto code that's removed by this PR?

I'm just really worried that this is going to break everything for me. After having to downgrade dask, use a patched conda, use a patched odo, and run boto master on Python until the last release, I'm super wary about these things.

**Comment by jreback** [comment]
this won't be till 0.19 I think we have it marked in any event

**Comment by TomAugspurger** [comment]
> Isn't there a try/except for that in the boto code that's removed by this PR?

Yep, I was hoping to avoid that in pandas, but it might be necessary.

@jseabold I have a bunch of stuff on AWS too and I share your concern about this breaking things.
I'll be writing many more tests. This is still a ways from being done, but I'll keep you updated.

**Comment by martindurant** [comment]
I'm not sure if we're dealing with one issue here or more: 
- the `profile_name` keyword to S3FileSystem is designed to be passed along to boto3.Session, e;
- connections are attempted with `anon=False` first, and then fallback to `True` if not specified;
- publicly available buckets should always be available.

I don't have boto profiles, just a [default] section in my .aws/credentials (no relevant env variables). How should I go about recreating your situation? In any case, I suggest the discussion should happen over on s3fs.

**Comment by jseabold** [comment]
@martindurant I've started today trying to collect the info for my typical permissions, so you can replicate this stuff.

**Comment by jseabold** [comment]
Is the original motivating bug in boto fixed with the newest release? I'm not sure if it's the same config parsing one I alluded to with conda that I fixed with a monkey patch.

What's the motivation for this over using boto3? I need to read the linked thread more carefully. I'd tentatively suggest not to add reliance on another third party package and introduce some more complexity here. The `read_csv` code is already pretty hard to follow with the conflation of source location with file format and all the options without having to dig through another package in addition to the Amazon ones.

**Comment by TomAugspurger** [comment]
> Is the original motivating bug in boto fixed with the newest release? I'm not sure if it's the same config parsing one I alluded to with conda that I fixed with a monkey patch.

Yep, not wanting to manage the boto to boto3 switch within pandas was what initially prompted this.  
It's a tradeoff between the complexity of (re)implementing it within pandas vs. the complexity of accepting a dependency. I'll try to get something more polished done tonight.

**Comment by jseabold** [comment]
> It's a tradeoff between the complexity of (re)implementing it within pandas vs. the complexity of accepting a dependency.

That's true, but you have control over what is getting implemented, so you don't have extra calls around things that a filesystem mimic might want, but a single stream a key to memory thing might not. To be fair, it looks like the `S3File` objects avoid some of the problems that I've been running in to with permissions.

This is probably out of scope for this PR, but I've often wanted to be able to pass a boto(3) `Key` object to `read_csv` rather than having to check how pandas is doing the session creation. If `S3File` is sufficiently file-like I'd love to skip the discovery and make me an s3 client steps, so I don't have to remind myself.

**Comment by jreback** [comment]
@jseabold this is to REDUCE complexity in the code. ``boto3`` is somewhat raw. This makes us just treat it directly as a filesystem and is pretty generic. This is how all file-ilke things should be done, abstract them away from pandas directly. (e.g.this is done implicity with compression / unicode conversions, though these should be abstracted a bit more, but they are usually just a line or 2). 

This makes it so pandas can simply use these w/o having to worry about messy details like authentication and such which the service can simply care about (and upgrade / fix as needed).

**Comment by martindurant** [comment]
You might be interested that https://github.com/dask/hdfs3 provides the very same interface to HDFS - perhaps a less used data source for pandas, but an option that might be useful in the future. We may also make the same interface for other cloud storage services, depending on demand/complexity.

**Comment by jseabold** [comment]
I agree with the sentiment and am all for reducing the complexity (in pandas) around csv reading, file-like objects, location discovery, and authentication around those locations.

**Comment by jseabold** [comment]
@TomAugspurger [This](https://github.com/dask/s3fs/pull/46#issuecomment-220775582) may be what's going on with  `anon`  for you.

**Comment by jreback** [comment]
update for 0.19.0?

File: pandas/io/tests/test_pickle.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 from pandas.tseries.offsets import Day, MonthEnd
 
 
-class TestPickle():
+class TestPickle(tm.TestCase):
```
Comment (sinhrks) [inline pandas/io/tests/test_pickle.py]: Does it run test generators? see docstring:

```
    NOTE: TestPickle can't be a subclass of tm.Testcase to use test generator.
    http://stackoverflow.com/questions/6689537/
    nose-test-generators-inside-class
```

Comment (TomAugspurger) [inline pandas/io/tests/test_pickle.py]: I think this was a stray commit from a different branch, trying to get pytest running. Will remove.

**Comment by TomAugspurger** [comment]
I'll revisit this mid next week.

**Comment by jreback** [comment]
@TomAugspurger ?

**Comment by jorisvandenbossche** [comment]
@TomAugspurger What's the status of this?

**Comment by TomAugspurger** [comment]
Pushing I think. Haven't had time to work on it, and there's one more change I want to submit to s3fs.

**Comment by jreback** [comment]
@TomAugspurger would be great for 0.20.0!

**Comment by TomAugspurger** [comment]
Pushed an update, will add a bit more for the docs later. Ended up being pretty simple.

**Comment by TomAugspurger** [comment]
Anecdotal performance boost:

```
# first one is master
bash-3.2$ nosetests pandas/io/tests/parser/test_network.py:TestS3
..........
----------------------------------------------------------------------
Ran 10 tests in 62.720s

OK
# now on this branch
bash-3.2$ git co -
Switched to branch 's3fs'
Your branch is up-to-date with 'origin/s3fs'.
bash-3.2$ nosetests pandas/io/tests/parser/test_network.py:TestS3
..........
----------------------------------------------------------------------
Ran 10 tests in 15.995s

OK
```

4x speedup seems a bit too good to be true :)

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -40,6 +40,12 @@
 import pandas.lib as lib
 import pandas.parser as _parser
 
+try:
```
Comment (jreback) [inline pandas/io/parsers.py]: move to io/common.py

**Comment by jreback** [comment]
let's do this after #14576 as that's changes the paths for the io code

**Comment by TomAugspurger** [comment]
I think this is ready to go, if anyone has further comments.

File: doc/source/whatsnew/v0.20.0.txt [hunk 4]
```diff
@@ -93,17 +93,15 @@ Backwards incompatible API changes
 
 .. _whatsnew_0200.api:
 
-
+- pandas now uses `s3fs <http://s3fs.readthedocs.io/>`_ for handling S3 connections. This shouldn't break
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: *maybe* make this a note / warning box?

Comment (TomAugspurger) [inline doc/source/whatsnew/v0.20.0.txt]: Warning looked a bit strange, so I made it a separate section (same level as "Other API Changes").

screen shot 2016-12-16 at 7 54 30 am

File: doc/source/whatsnew/v0.20.0.txt [hunk 5]
```diff
@@ -97,13 +97,16 @@ Backwards incompatible API changes
 - ``CParserError`` has been renamed to ``ParserError`` in ``pd.read_csv`` and will be removed in the future (:issue:`12665`)
 - ``SparseArray.cumsum()`` and ``SparseSeries.cumsum()`` will now always return ``SparseArray`` and ``SparseSeries`` respectively (:issue:`12855`)
 
+S3 File Handling
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: maybe add a ref tag

File: doc/source/whatsnew/v0.20.0.txt [hunk 6]
```diff
@@ -97,13 +97,16 @@ Backwards incompatible API changes
...
+^^^^^^^^^^^^^^^^
 
-
+pandas now uses `s3fs <http://s3fs.readthedocs.io/>`_ for handling S3 connections. This shouldn't break
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: do we need a doc section itself somewhere? so that we can put a url in various doc-strings, e.g. read_csv? (can always do as a followup)

**Comment by martindurant** [comment]
No comments from me on the pandas code.
I *do* expect potential problems in the future as boto3/botocore changes internal things, but that would have been true for any wrapping. Will need to keep an eye on that.

File: doc/source/install.rst [hunk 7]
```diff
@@ -262,7 +262,7 @@ Optional Dependencies
   * `XlsxWriter <https://pypi.python.org/pypi/XlsxWriter>`__: Alternative Excel writer
 
 * `Jinja2 <http://jinja.pocoo.org/>`__: Template engine for conditional HTML formatting.
-* `boto <https://pypi.python.org/pypi/boto>`__: necessary for Amazon S3 access.
+* `s3fs <http://s3fs.readthedocs.io/>`__: necessary for Amazon S3 access.
```
Comment (jreback) [inline doc/source/install.rst]: do we have a min requirement for s3fs? (0.7.0)?

Comment (TomAugspurger) [inline doc/source/install.rst]: Not yet, may as well require the latest? (`0.0.7`)

Comment (jreback) [inline doc/source/install.rst]: I *think* older versions work though, so maybe this is not an issue.

**Comment by dhimmel** [comment]
It probably makes sense to merge https://github.com/pandas-dev/pandas/pull/14880 before this, which is part 2 of the compression refactoring. https://github.com/pandas-dev/pandas/pull/14880 changes a few of the S3 tests.

**Comment by jreback** [comment]
@TomAugspurger give this a rebase and test.

**Comment by TomAugspurger** [comment]
Travis is green.

**Comment by jreback** [comment]
thanks @TomAugspurger 

I changed a name in the excel tests.
